File: buffer.py
import threading

class Buffer:
    def __init__(self):
        self.data = None
        self.lock = threading.Lock()
    
    def add(self, msg):
        with self.lock:
            self.data = msg

    def get(self):
        with self.lock:
            msg = self.data
            self.data = None
            return msg
        
    def isEmpty(self):
        return self.data is None


if __name__ == "__main__":
    buffer = Buffer()
    print(buffer.get())

    buffer.add(1)
    buffer.add("Fist")
    buffer.add([1,2,3])
    item = buffer.get()
    print(item)

File: C++ test code/buffer.cpp
#include <iostream>
#include <string>
#include <mutex>
#include "buffer.h"

using namespace std;

Buffer::Buffer() {
    data = "";
}

void Buffer::add(string msg) {
    mtx.lock();
    data = msg;
    mtx.unlock();
}

string Buffer::get() {
    mtx.lock();
    string msg = data;
    data = "";
    mtx.unlock();
    return msg;
}

bool Buffer::isEmpty() {
    // cout << "debugging isEmpty data: " << data << ", length: " << data.length() << endl;
    return data.length() == 0;
}


// int main(int argc, char* argv[]) {
//     Buffer buffer;
//     cout << buffer.get() << endl;

//     buffer.add("V pos");
//     buffer.add("Fist");
//     buffer.add("1, 2, 3");
//     string item = buffer.get();
//     cout << item << endl;
// }